From 1086cdb8fd77a224d56033bde0825a286ba30ee2 Mon Sep 17 00:00:00 2001
From: Vincent Bernat <vincent@bernat.ch>
Date: Wed, 22 Aug 2018 19:20:35 +0200
Subject: [PATCH] use SNI when connecting with SSL

imap.gmail.com doesn't accept connections without SNI anymore. Without
this extension, it returns a self-signed certificate and mbsync is
unable to complete:

    $ openssl s_client -connect imap.gmail.com:993 -noservername
    CONNECTED(00000005)
    depth=0 OU = "No SNI provided; please fix your client.", CN = invalid2.invalid
    verify error:num=18:self signed certificate
    verify return:1
    depth=0 OU = "No SNI provided; please fix your client.", CN = invalid2.invalid
    verify return:1
    ---
    Certificate chain
     0 s:OU = "No SNI provided; please fix your client.", CN = invalid2.invalid
       i:OU = "No SNI provided; please fix your client.", CN = invalid2.invalid

This commit configure the SSL connection to transmit the hostname
through SNI. This has been tested with both GMail (which requires SNI)
and Fastmail (which doesn't require SNI).
---
 src/socket.c | 1 +
 1 file changed, 1 insertion(+)

--- a/src/socket.c
+++ b/src/socket.c
@@ -270,6 +270,7 @@
 
 	init_wakeup( &conn->ssl_fake, ssl_fake_cb, conn );
 	conn->ssl = SSL_new( ((server_conf_t *)conn->conf)->SSLContext );
+	SSL_set_tlsext_host_name( conn->ssl, conn->conf->host );
 	SSL_set_fd( conn->ssl, conn->fd );
 	SSL_set_mode( conn->ssl, SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER );
 	socket_expect_read( conn, 1 );
